# Bài tập 2

Bùi Văn Hợp - 1712046

## Câu 1 (1 điểm)

In [1]:
import numpy as np

def hoeffding_inequality(nu, mu, N, eps):
    left_ine = 1 if (abs(nu - mu) > eps) else 0
    right_ine = 2 * np.exp(-2 * eps**2 * N)
    return (left_ine <= right_ine)

def hoeffding_check(nu, mu, N):
    eps_range = np.arange(0.01, 0.51, 0.01)
    for eps in eps_range:
        if (hoeffding_inequality(nu, mu, N, eps) == False):
            return False
    return True

def run_stimulation(n_test, n_coin, n_times):
    nus = np.random.binomial(n_times, 0.5, (n_test, n_coin)) / n_times
    v_1 = np.mean(nus[:, 0])
    pos_rand = np.random.choice(n_coin, size=n_test)
    v_rand = np.mean(nus[np.arange(n_test), pos_rand])
    v_min = np.mean(np.min(nus, axis=1))
    print("v_1 = ", v_1)
    print("v_rand = ", v_rand)
    print("v_)min = ", v_min)
    print("Checking Hoeffding inequality for v_1 : ", hoeffding_check(v_1, 0.5, 10))
    print("Checking Hoeffding inequality for v_rand : ", hoeffding_check(v_rand, 0.5, 10))
    print("Checking Hoeffding inequality for v_min : ", hoeffding_check(v_min, 0.5, 10))


In [2]:
run_stimulation(n_test=100000, n_coin=1000, n_times=10)

v_1 =  0.5002099999999998
v_rand =  0.500507
v_)min =  0.037512
Checking Hoeffding inequality for v_1 :  True
Checking Hoeffding inequality for v_rand :  True
Checking Hoeffding inequality for v_min :  False


Từ kết quả chạy giả lập, $\nu_{min}$ xấp xỉ khoảng 0.0375, nên mình **chọn đáp án là b, 0.01**.

YOUR ANSWER HERE b

## Câu 2 (1 điểm)

Trong hàm chạy giả lập, mình cho thử bất đẳng thức Hoeffding với khoảng các giá trị $\epsilon$ trong khoảng từ 0.01 đến 0.5. Kết quả nhận được là $\nu_1$ và $\nu_{rand}$ tuân theo bất đẳng thức Hoeffding, còn $\nu_{min}$ không tuân theo bất đẳng thức. Nên mình sẽ chọn đáp án **d, c1 và crand**

**Giải thích thêm về toán học:**

- Ta biết rằng mỗi đồng xu trong 1000 đồng xu đều độc lập với nhau, và việc ta tung mỗi đồng xu 10 lần cũng là những sự kiện độc lập với nhau.

- Vậy nên $c_1$ (có xác suất ra mặt sấp là $\nu_1$) là cách ta chọn đồng xu thứ nhất (duy nhất 1 đồng xu) trong 1000 đồng xu, và tung nó 10 lần, nên nó sẽ luôn tuân theo bất đẳng thức Hoeffding (cho 1 bin duy nhất).

- Với $c_{rand}$ (có xác suất ra mặt sấp là $\nu{rand}$), ở trên ta đã biết mỗi đồng xu là độc lập với nhau, nên việc trong 10000 lần, ta chọn ngẫu nhiên 1 đồng xu, cũng giống như việc chọn 1 đồng xu cố định. Nên $\nu_{rand}$ cũng sẽ phù hợp với bất đẳng thức.

- Riêng $c_{min}$, vì ta có chủ đích chọn đồng xu có xác suất ra mặt sấp trong 10 lần nhỏ nhất, nên nó sẽ theo 1 phân phối khác so với cách chọn đồng xu trên. 

**Ta sẽ hiểu bài toán này theo 1 cách khác như sau,**, mỗi đồng xu là 1 bin, và việc tung đồng xu 10 lần là cách chúng ta lấy 10 sample trong bin đó. Và chúng ta sẽ có 1000 bin.

Giả sử ta cần tìm $P(\nu_{min}) = 0$, với $\mu_{min} = 0.5$ (Vì đây là 1 đồng xu đồng chất xác suất sấp-ngửa bằng nhau).

Với bài tập lần trước, ta dễ dàng tính ra $P(\nu_{min}) = 1 - (1 - 0.5^{10})^{1000} \approx 0.62$

Ta có bất đẳng thức Hoeffding áp dụng cho đồng xu này như sau:

$P(|\nu_{min} - \mu_{min}| > \epsilon) \leq 2e^{-2\epsilon^2 N}$

Ta đặt $\epsilon = 0.49$, ta có $P(|\nu_{min} - \mu_{min}| > \epsilon) = P(|0 - 0.5| > 0.49)$

$\Rightarrow P(\nu_{min}) = 0.62 \leq 2e^{-2\epsilon^2 N} \leq 0.016$ **(vô lý)**

Nên đồng xu $c_{min}$ sẽ không tuân theo bất đẳng thức Hoeffding.


YOUR ANSWER HERE d

## Câu 3 (1 điểm)

Từ bài toán ta biết được xác suất để hypothesis $h$ sai so với target function $f(x)$ là $\mu$

Vậy ta có: $P(h(x) \neq f(x)) = \mu, P(h(x) = f(x)) = 1 - \mu$

Với nhãn $y$ có một độ lỗi nhất định so với target function $f(x)$, ta suy ra từ bài:

$P(y = f(x)) = \lambda, P(y \neq f(x)) = 1 - \lambda$

Ta cần tìm giá trị của $P(h(x) \neq y)$ , ta sẽ sử dụng xác suất điều kiện với 2 điều kiện như sau:

- Nếu $h(x) \neq f(x)$, thì ta cần $y = f(x)$ để thỏa mãn điều kiện $h(x) \neq y$.

- Nếu $h(x) = f(x)$, thì ta cần điều kiện $y \neq f(x)$ để thỏa mãn điều kiện $h(x) \neq y$.

Từ đó ta có, $P(h(x) \neq y) = P(h(x) \neq f(x) | y = f(x)) \times P(y = f(x) + P(h(x) = f(x) | y \neq f(x)) \times P(y \neq f(x))$

 $P(h(x) \neq y) = \mu\lambda + (1 - \mu)(1 - \lambda)$
 
 **Nên mình sẽ lựa chọn đáp án e**

YOUR ANSWER HERE e

## Câu 4 (1 điểm)

Từ kết quả của câu trên, ta có  $P(h(x) \neq y) = \mu\lambda + (1 - \mu)(1 - \lambda) = 2\mu\lambda - \mu - 1 - \lambda$

$P(h(x) \neq y) = \mu(2\lambda - 1) + 1 - \lambda$

Vậy để $P(h(x) \neq y)$ không phụ thuộc vào $\mu$, ta cần có $2\lambda - 1 = 0 \Rightarrow \lambda = 0.5$

**Vậy mình sẽ chọn đáp án b, 0.5.**

YOUR ANSWER HERE b

## Câu 5 (1 điểm)

In [3]:
def generate_target_w():
    """
    Generates target_w (the vector of parameters of f) 
    from two random, uniformly distributed points in [-1, 1] x [-1, 1].
    
    Returns
    -------
    target_w : numpy array, shape (3, 1) 
        The vector of parameters of f.
    """
    # Generate two points from a uniform distribution over [-1, 1]x[-1, 1]
    p1 = np.random.uniform(-1, 1, 2)
    p2 = np.random.uniform(-1, 1, 2)
    # Compute the target W from these two points
    target_w = np.array([p1[1]*p2[0] - p1[0]*p2[1], p2[1] - p1[1], p1[0] - p2[0]]).reshape((-1, 1))
    
    return target_w

In [4]:
def generate_data(N, target_w):
    """
    Generates a data set by generating random inputs and then using target_w to generate the 
    corresponding outputs.
    
    Parameters
    ----------
    N : int
        The number of examples.
    target_w : numpy array, shape (3, 1) 
        The vector of parameters of f.
    
    Returns
    -------
    X : numpy array, shape (N, 3)
        The matrix of input vectors (each row corresponds to an input vector); the first column of 
        this matrix is all ones.
    Y : numpy array, shape (N, 1)
        The vector of outputs.        
    """
    bad_data = True # `bad_data = True` means: data contain points on the target line 
                    # (this rarely happens, but just to be careful)
                    # -> y's of these points = 0 (with np.sign); 
                    #    we don't want this (y's of data must be -1 or 1)
                    # -> re-generate data until `bad_data = False`
    
    while bad_data == True:
        X = np.random.uniform(-1, 1, (N, 2))
        X = np.hstack((np.ones((N, 1)), X)) # Add 'ones' column
        Y = np.sign(np.dot(X, target_w))
        if (0 not in Y): # Good data
            bad_data = False
    
    return X, Y

In [5]:
def run_linear_regression(X, Y):
    X_dagger = np.dot(np.linalg.pinv(np.dot(X.T, X)), X.T)
    w = np.dot(X_dagger, Y)
    return w

In [9]:
def main(N):
    num_run = 1000
    avg_error_in = 0.0
    avg_error_out = 0.0
    for _ in range(num_run):
        target_w = generate_target_w()
        X, Y = generate_data(N, target_w)
        w = run_linear_regression(X, Y)
        
        error_in = np.mean(np.sign(np.dot(X, w)) != Y)
        avg_error_in += error_in
        
        X_test, Y_test = generate_data(1000, target_w)
        error_out = np.mean(np.sign(np.dot(X_test, w)) != Y_test)
        avg_error_out += error_out
    
    avg_error_in /= num_run
    avg_error_out /= num_run
    print("Avg Error in : ", avg_error_in)
    print("Avg Error out : ", avg_error_out)

In [11]:
main(N=100)

Avg Error in :  0.03848000000000006
Avg Error out :  0.046965000000000055


Từ kết quả Độ lỗi trung bình của tập train (Error_in) khoảng 0.03, **nên mình sẽ chọn đáp án gần nhất là c, 0.01**

YOUR ANSWER HERE c

## Câu 6 (1 điểm)

Cũng với hàm trên, Độ lỗi trung bình của tập test (Error_out) lớn hơn Error_in một chút, những cũng xấp xỉ 0.04, **nên mình chọn đáp án gần nhất cũng là c, 0.01**

YOUR ANSWER HERE c

## Câu 7 (1 điểm)

In [13]:
def run_PLA(X, Y, w):
    iteration = 0
    while True:
        Y_hat = np.sign(X.dot(w))
        x_pos, y_pos = np.where(Y_hat != Y)
        if len(x_pos) == 0:
            break
        random_x = np.random.choice(x_pos)
        w = w + Y[random_x, 0] * X[random_x, :].reshape(-1, 1)
        iteration += 1
          
    return w, iteration

In [15]:
def main_PLA_linear_regression(N):
    num_run = 1000
    avg_error_in = 0.0
    avg_error_out = 0.0
    avg_iteration = 0.0
    for _ in range(num_run):
        target_w = generate_target_w()
        X, Y = generate_data(N, target_w)
        w = run_linear_regression(X, Y)
        w, iteration = run_PLA(X, Y, w)
        
        avg_iteration += iteration
        
        error_in = np.mean(np.sign(np.dot(X, w)) != Y)
        avg_error_in += error_in
        
        X_test, Y_test = generate_data(1000, target_w)
        error_out = np.mean(np.sign(np.dot(X_test, w)) != Y_test)
        avg_error_out += error_out
    
    avg_error_in /= num_run
    avg_error_out /= num_run
    avg_iteration /= num_run
    print("Avg Error in : ", avg_error_in)
    print("Avg Error out : ", avg_error_out)
    print("Avg Iteration: ", avg_iteration)

In [19]:
main_PLA_linear_regression(N=10)

Avg Error in :  0.0
Avg Error out :  0.09812799999999997
Avg Iteration:  4.366


Sau khi áp dụng Linear Regression và tiếp tục cho vào thuật toán PLA để tìm ra hypothesis thì trung bình số lần lặp là khoảng 4 -> 5. **Nên mình sẽ chọn đáp án gần nhất là a, 1.**

YOUR ANSWER HERE a

## Câu 8 (1 điểm)

YOUR ANSWER HERE

## Câu 9 (1 điểm)

YOUR ANSWER HERE

## Câu 10 (1 điểm)

YOUR ANSWER HERE